src/fftr.nim

```diff
@@ -323,25 +323,105 @@ func process*(
+  ## - normalize: if true, normalizes the results by dividing by the length of the signal when inverse is true.
+  ## - n: Length of the the output.
+  ##      If `n` is not given, the length of the input is automatically used.
+  ##      If `n` is smaller than the length of the input, the input is cropped.
+  ##      If `n` is larger, the input is padded with zeros.
+  let size = if n <= 0: input.len else: n
+
+  if size < input.len:
+    # Extend the input with zeros as much as needed
+    let input = input.toSeq & newSeq[Complex64](size - input.len)
+  elif size < input.len:
+    # Crop the input as much as needed
+    let input = input[0..<size]
+
+  dftImpl(input, inverse=inverse, normalize=normalize)
+
+func fftImpl(input: openArray[Complex64], inverse: bool = false, normalize: bool = false): seq[Complex64] =
+  ## Calculates the FFT or the IFFT of an input signal using the best method given the input length
+  ##
+  ## Inputs:
+  ## - input: The input signal
+  ## - inverse: if true, calculates the inverse FFT, otherwise calculates the FFT
+  ## - normalize: if true, normalizes the results by dividing by the length of the signal when inverse is true
+  let size = input.len
+  result.setLen(size)
+
+  if isPowerOfTwo(size):
+    var ctx = Stockham2.init(size, inverse)
     process(ctx, input, result)
-  elif input.len.countTrailingZeroBits == 0:
-    var ctx = Bluestein.init(input.len, inverse)
+  elif size.countTrailingZeroBits == 0:
+    var ctx = Bluestein.init(size, inverse)
     process(ctx, input, result)
   else:
-    var ctx = RadixMixed.init(input.len, inverse)
+    var ctx = RadixMixed.init(size, inverse)
     process(ctx, input, result)
+
+  if inverse and normalize:
+    let complexInputLen = complex64(float64(size))
+    for i in 0..<result.len:
+      result[i] = result[i] / complexInputLen
+
+func fft*(input: openArray[Complex64], inverse: bool = false, normalize: bool = false, n = 0): seq[Complex64] =
+  ## Calculates the FFT or the IFFT of an input signal
+  ##
+  ## This function automatically selects the best calculation method given the desired FFT size
+  ##
+  ## Inputs:
+  ## - input: The input signal.
+  ## - inverse: if true, calculates the inverse FFT, otherwise calculates the FFT (which is the default).
+  ## - normalize: if true, normalizes the results by dividing by the length of the signal when inverse is true.
+  ## - n: Length of the the output.
+  ##      If `n` is not given, the length of the input is automatically used.
+  ##      If `n` is smaller than the length of the input, the input is cropped.
+  ##      If `n` is larger, the input is padded with zeros.
+  let size = if n <= 0: input.len else: n
+
+  let input = if size > input.len:
+    # Extend the input with zeros as much as needed
+    input.toSeq & newSeq[Complex64](size - input.len)
+  elif size < input.len:
+    # Crop the input as much as needed
+    input[0..<size]
+  else:
+    input.toSeq
```

Comment:
and another copy here - adding these copies up would significantly slow down the implementation - you should see this if you run the benchmark before and after

Reply:
FWIW, I ran the benchmarks and I did not notice any significant differences, but I agree that unnecessary copies should be avoided.

Reply:
I've improved the code to avoid unnecessary copies.